### DISCUSSION TITLE
Why does `gl` not select text visually?

### DISCUSSION THREAD

**Discussion by txtyash** [helix-editor/helix#1215] [Ideas]
`gl` doesn't select text till the end and neither does `gh` till the start. That way deleting text till the end or till the beginning could be one key stroke shorter as opposed to `vgld` or `vghd`. Even `ge` should visually select the text till the end so that `ged` is a better alternative to `vxgegld`.

**Comment by NNBnh** [comment]
Yeah, this make all movement action consistent with _find_, _till_, _word-next_... and I don't see any downside of implement this. All [goto movements](https://docs.helix-editor.com/keymap.html#goto-mode), [unimpaired movements](https://docs.helix-editor.com/keymap.html#unimpaired) and [`match_brackets`](https://docs.helix-editor.com/keymap.html#match-mode) should also select (similar to Kakoune).

**Reply by pickfire** [reply]
Maybe can submit a pull request for this. I did mentioned this in the chat before as well.

**Reply by archseer** [reply]
> (similar to Kakoune)

Note that this is actually under `G` in kakoune, not `g`.

Yeah I'm undecided here. I think it could be a good change, but at the same time not all commands can do this (for example `gd`, `gf` etc.) so it would be a bit inconsistent. It makes sense for `gl`, `gh` etc. but when using `gg`/`ge` I tend to use them for jumping, not selecting. Not sure.

**Reply by NNBnh** [reply]
> Not all commands can do this (for example `gd`, `gf` etc.) so it would be a bit inconsistent.

Don't worry, it's not inconsistent at all! It's impossible to select from one file to another, and `goto_file` is more of a buffer-action rather than a movement to begin with ;-).

> When using `gg`/`ge` I tend to use them for jumping, not selecting.

In Kakoune i sometime delete everything from the current position to the EOF to quickly rewrite the last `void main() { ... }` function inside my file. This proves that although not often, it have it's use case.

**Reply by sourcefrog** [reply]
I also notice this a lot with the home/end keys, which I suppose are bound to the same action.

**Comment by sudormrfbin** [comment]
It can be configured to extend instead of move, though I agree that it makes more sense for it to extend by default (even though inconsistent). I guess it's similar to the age-old `Y` vs `y$` debate in vim :)

```toml
[keys.normal]
"g" = { "l" = "extend_to_line_end", "h" = "extend_to_line_start" }
```

**Reply by txtyash** [reply]
> but when using gg/ge I tend to use them for jumping, not selecting.

@archseer don't you think both actions could happen with the same keybinding? like `ge`
takes your cursor to the end while also selecting the text till the end of the
file?

**Reply by pickfire** [reply]
From the chat from what I know the people do agree that extending visually for goto commands as the default is good too. Just that up till now no one submitted a pull request for it yet. Maybe I might submit a pull request when I get annoyed enough to create a pull request, soon maybe.

**Reply by rkshthrmsh** [reply]
Thanks for the suggestion. Is there a way to immitate this extend behavior for end of file? Something like:
```toml
[keys.normal]
"g" = { "e" = "extend_to_file_end" }
```

Could not find anything relate to "extend_to_file_end" in the docs.

EDIT:
```toml
[keys.normal]
g = { e = ["select_mode", "goto_file_end", "exit_select_mode"]  }
 ```

**Reply by rcorre** [reply]
If I have
abc `def`| ghi

Then I'd expect `gh` to give me the same as `T<cr>`, i.e.:
|`abc def` ghi

Instead I get
|`abc d`ef ghi

The same is true of the `gs` binding suggested by https://github.com/helix-editor/helix/discussions/1215#discussioncomment-5619893 (which we don't have a good `t` equivalent for).

**Comment by txtyash** [comment]
If selections are being avoided with these movements due to ugliness reasons then there could be another way UI wise to represent the selection without making the text look ugly.

**Comment by rcorre** [comment]
By default, I think`t<cr>` and `T<cr>` have the behavior you're looking for. I use "to the newline" as a mnemonic

**Reply by rcorre** [reply]
I'm realizing `t<cr>`/`T<cr>` will not work on the first line, or the last if the file doesn't end with a newline, so these are not an ideal solution.

**Comment by rkshthrmsh** [comment]
Apart from `gh` and `gl` for selecting till the start of line and end of line respectively, it might also be useful to have `gs` select up to the first non-blank in line.

**Reply by rkshthrmsh** [reply]
One keymap that works for this:
```toml
[keys.normal]
g = { s = ["select_mode", "goto_first_nonwhitespace", "exit_select_mode"] }
```

**Reply by rcorre** [reply]
I ended up with this, to avoid the issue mentioned at https://github.com/helix-editor/helix/discussions/1215#discussioncomment-5620016:

```
s = ["collapse_selection", "select_mode", "goto_first_nonwhitespace", "exit_select_mode"]
```
